packages/remix-cloudflare-workers-esm/worker.ts

```diff
@@ -0,0 +1,174 @@
+type WaitUntilCallback = KvAssetHandlerEvent["waitUntil"];
+
+const module = JSON.parse(moduleJSON) as AssetManifest;
+
+/**
+ * A function that returns the value to use as `context` in route `loader` and
+ * `action` functions.
+ *
+ * You can think of this as an escape hatch that allows you to pass
+ * environment/platform-specific values through to your loader/action.
+ */
+export type GetLoadContextFunction<Env = unknown> = (
+  request: Request,
+  env: Env,
+  ctx: ExecutionContext
+) => AppLoadContext;
+
+export type RequestHandler = ReturnType<typeof createRequestHandler>;
+
+/**
+ * Returns a request handler for the Cloudflare runtime that serves the
+ * Remix SSR response.
+ */
+export function createRequestHandler<Env = unknown>({
+  build,
+  getLoadContext,
+  mode,
+}: {
+  build: ServerBuild;
+  getLoadContext?: GetLoadContextFunction<Env>;
+  mode?: string;
+}) {
+  let handleRequest = createRemixRequestHandler(build, mode);
+
+  return (request: Request, env: Env, ctx: ExecutionContext) => {
+    let loadContext = getLoadContext?.(request, env, ctx);
+
+    return handleRequest(request, loadContext);
+  };
+}
+
+export async function handleAsset(
+  request: Request,
+  waitUntil: WaitUntilCallback,
+  assetNamespace: KVNamespace<string>,
+  build: ServerBuild,
+  options?: Partial<KvAssetHandlerOptions>
+) {
+  try {
+    let mergedOptions = {
+      ASSET_NAMESPACE: assetNamespace,
+      ASSET_MANIFEST: module,
+      ...options,
+    };
+
+    if (process.env.NODE_ENV === "development") {
+      return await getAssetFromKV(
+        { request, waitUntil },
+        {
+          cacheControl: {
+            bypassCache: true,
+          },
+          ...mergedOptions,
+        }
+      );
+    }
+
+    let cacheControl = {};
+    let url = new URL(request.url);
+    let assetpath = build.assets.url.split("/").slice(0, -1).join("/");
+    let requestpath = url.pathname.split("/").slice(0, -1).join("/");
+
+    if (requestpath.startsWith(assetpath)) {
+      // Assets are hashed by Remix so are safe to cache in the browser
+      // And they're also hashed in KV storage, so are safe to cache on the edge
+      cacheControl = {
+        bypassCache: false,
+        edgeTTL: 31536000,
+        browserTTL: 31536000,
+      };
+    } else {
+      // Assets are not necessarily hashed in the request URL, so we cannot cache in the browser
+      // But they are hashed in KV storage, so we can cache on the edge
+      cacheControl = {
+        bypassCache: false,
+        edgeTTL: 31536000,
+      };
+    }
+
+    return await getAssetFromKV(
+      { request, waitUntil },
+      {
+        cacheControl,
+        ...mergedOptions,
+      }
+    );
+  } catch (error) {
+    if (
+      error instanceof MethodNotAllowedError ||
+      error instanceof NotFoundError
+    ) {
+      return null;
+    }
+
+    throw error;
+  }
+}
+
+export function createEventHandler<Env = unknown>({
```

Comment:
I’ve parameterised `Env` here to allow the developer to supply their own contract as necessary. Better than trying to override it.